src/Identifier/PasswordIdentifier.php

```diff
@@ -88,61 +88,58 @@ public function getPasswordHasher()
      */
     public function identify(array $data)
     {
-        $fields = $this->getConfig('fields');
-
-        $usernameFields = (array)$fields['username'];
-        $found = array_intersect_key(array_flip($usernameFields), $data);
-        if (empty($found)) {
+        if (!isset($data['username'])) {
             return null;
         }
 
-        $password = null;
-        if (!empty($data[$fields['password']])) {
-            $password = $data[$fields['password']];
+        $identity = $this->_findIdentity($data['username']);
+        if (array_key_exists('password', $data)) {
+            $password = $data['password'];
+            if (!$this->_checkPassword($identity, $password)) {
+                return null;
+            }
         }
 
-        return $this->_findIdentity($data[key($found)], $password);
+        return $identity;
     }
 
     /**
      * Find a user record using the username and password provided.
      * Input passwords will be hashed even when a user doesn't exist. This
      * helps mitigate timing attacks that are attempting to find valid usernames.
      *
-     * @param string $identifier The username/identifier.
-     * @param string|null $password The password, if not provided password checking is skipped
-     *   and result of find is returned.
-     * @return \ArrayAccess|null User data entity or null on failure.
+     * @param array|\ArrayAccess|null $identity The identity or null.
+     * @param string|null $password The password.
+     * @return bool
      */
-    protected function _findIdentity($identifier, $password = null)
+    protected function _checkPassword($identity, $password)
     {
-        $result = $this->_findUser($identifier);
-        if (empty($result)) {
-            return null;
+        if ($identity === null) {
+            $identity = [
+                'password' => ''
+            ];
         }
 
-        if ($password !== null) {
-            $passwordField = $this->getConfig('fields.password');
-            $hasher = $this->getPasswordHasher();
-            $hashedPassword = $result[$passwordField];
-            if (!$hasher->check($password, $hashedPassword)) {
-                return null;
-            }
-
-            $this->_needsPasswordRehash = $hasher->needsRehash($hashedPassword);
-            unset($result[$passwordField]);
+        $passwordField = $this->getConfig('fields.password');
+        $hasher = $this->getPasswordHasher();
+        $hashedPassword = $identity[$passwordField];
```

Comment:
We would still need a `passwordField` configuration option if we were to make an 'interface' as I had proposed. :cry: